File: cadence/lib/go/test/kibble_test.go
package test

import (
	"regexp"
	"testing"

	"github.com/onflow/cadence"
	jsoncdc "github.com/onflow/cadence/encoding/json"
	emulator "github.com/onflow/flow-emulator"
	"github.com/onflow/flow-go-sdk"
	sdk "github.com/onflow/flow-go-sdk"
	"github.com/onflow/flow-go-sdk/crypto"
	"github.com/onflow/flow-go-sdk/templates"
	"github.com/onflow/flow-go-sdk/test"
	"github.com/stretchr/testify/assert"

	ft_contracts "github.com/onflow/flow-ft/lib/go/contracts"
)

const (
	kibbleRootPath           = "../../.."
	kibbleContractPath       = kibbleRootPath + "/contracts/Kibble.cdc"
	kibbleSetupAccountPath   = kibbleRootPath + "/transactions/kibble/setup_account.cdc"
	kibbleTransferTokensPath = kibbleRootPath + "/transactions/kibble/transfer_tokens.cdc"
	kibbleMintTokensPath     = kibbleRootPath + "/transactions/kibble/mint_tokens.cdc"
	kibbleBurnTokensPath     = kibbleRootPath + "/transactions/kibble/burn_tokens.cdc"
	kibbleGetBalancePath     = kibbleRootPath + "/scripts/kibble/get_balance.cdc"
	kibbleGetSupplyPath      = kibbleRootPath + "/scripts/kibble/get_supply.cdc"
)

func KibbleDeployContracts(b *emulator.Blockchain, t *testing.T) (flow.Address, flow.Address, crypto.Signer) {
	accountKeys := test.AccountKeyGenerator()

	// Should be able to deploy a contract as a new account with no keys.
	fungibleTokenCode := loadFungibleToken()
	fungibleAddr, err := b.CreateAccount(
		[]*flow.AccountKey{},
		[]templates.Contract{templates.Contract{
			Name:   "FungibleToken",
			Source: string(fungibleTokenCode),
		}},
	)
	assert.NoError(t, err)

	_, err = b.CommitBlock()
	assert.NoError(t, err)

	kibbleAccountKey, kibbleSigner := accountKeys.NewWithSigner()
	kibbleCode := loadKibble(fungibleAddr)

	kibbleAddr, err := b.CreateAccount(
		[]*flow.AccountKey{kibbleAccountKey},
		[]templates.Contract{templates.Contract{
			Name:   "Kibble",
			Source: string(kibbleCode),
		}},
	)
	assert.NoError(t, err)

	_, err = b.CommitBlock()
	assert.NoError(t, err)

	// Simplify testing by having the contract address also be our initial Vault.
	KibbleSetupAccount(t, b, kibbleAddr, kibbleSigner, fungibleAddr, kibbleAddr)

	return fungibleAddr, kibbleAddr, kibbleSigner
}

func KibbleSetupAccount(t *testing.T, b *emulator.Blockchain, userAddress sdk.Address, userSigner crypto.Signer, fungibleAddr sdk.Address, kibbleAddr sdk.Address) {
	tx := flow.NewTransaction().
		SetScript(kibbleGenerateSetupKibbleAccountTransaction(fungibleAddr, kibbleAddr)).
		SetGasLimit(100).
		SetProposalKey(b.ServiceKey().Address, b.ServiceKey().Index, b.ServiceKey().SequenceNumber).
		SetPayer(b.ServiceKey().Address).
		AddAuthorizer(userAddress)

	signAndSubmit(
		t, b, tx,
		[]flow.Address{b.ServiceKey().Address, userAddress},
		[]crypto.Signer{b.ServiceKey().Signer(), userSigner},
		false,
	)
}

func KibbleCreateAccount(t *testing.T, b *emulator.Blockchain, fungibleAddr sdk.Address, kibbleAddr sdk.Address) (sdk.Address, crypto.Signer) {
	userAddress, userSigner, _ := createAccount(t, b)
	KibbleSetupAccount(t, b, userAddress, userSigner, fungibleAddr, kibbleAddr)
	return userAddress, userSigner
}

func KibbleMint(t *testing.T, b *emulator.Blockchain, fungibleAddr sdk.Address, kibbleAddr sdk.Address, kibbleSigner crypto.Signer, recipientAddress flow.Address, amount string, shouldRevert bool) {
	tx := flow.NewTransaction().
		SetScript(kibbleGenerateMintKibbleTransaction(fungibleAddr, kibbleAddr)).
		SetGasLimit(100).
		SetProposalKey(b.ServiceKey().Address, b.ServiceKey().Index, b.ServiceKey().SequenceNumber).
		SetPayer(b.ServiceKey().Address).
		AddAuthorizer(kibbleAddr)

	_ = tx.AddArgument(cadence.NewAddress(recipientAddress))
	_ = tx.AddArgument(CadenceUFix64(amount))

	signAndSubmit(
		t, b, tx,
		[]flow.Address{b.ServiceKey().Address, kibbleAddr},
		[]crypto.Signer{b.ServiceKey().Signer(), kibbleSigner},
		shouldRevert,
	)

}

func TestKibbleDeployment(t *testing.T) {
	b := newEmulator()

	fungibleAddr, kibbleAddr, _ := KibbleDeployContracts(b, t)

	t.Run("Should have initialized Supply field correctly", func(t *testing.T) {
		supply := executeScriptAndCheck(t, b, kibbleGenerateGetSupplyScript(fungibleAddr, kibbleAddr), nil)
		assert.EqualValues(t, CadenceUFix64("0.0"), supply)
	})
}

func TestKibbleSetupAccount(t *testing.T) {
	b := newEmulator()

	t.Run("Should be able to create empty vault that does not affect supply", func(t *testing.T) {
		fungibleAddr, kibbleAddr, _ := KibbleDeployContracts(b, t)

		userAddress, _ := KibbleCreateAccount(t, b, fungibleAddr, kibbleAddr)

		userBalance := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr),
			[][]byte{jsoncdc.MustEncode(cadence.Address(userAddress))},
		)
		assert.EqualValues(t, CadenceUFix64("0.0"), userBalance)

		supply := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetSupplyScript(fungibleAddr, kibbleAddr),
			nil,
		)
		assert.EqualValues(t, CadenceUFix64("0.0"), supply)
	})
}

func TestKibbleMinting(t *testing.T) {
	b := newEmulator()

	fungibleAddr, kibbleAddr, kibbleSigner := KibbleDeployContracts(b, t)

	userAddress, _ := KibbleCreateAccount(t, b, fungibleAddr, kibbleAddr)

	t.Run("Should not be able to mint zero tokens", func(t *testing.T) {
		KibbleMint(t, b, fungibleAddr, kibbleAddr, kibbleSigner, userAddress, "0.0", true)
	})

	t.Run("Should be able to mint tokens, deposit, update balance and total supply", func(t *testing.T) {
		KibbleMint(t, b, fungibleAddr, kibbleAddr, kibbleSigner, userAddress, "50.0", false)

		// Assert that vault balance is correct
		userBalance := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr),
			[][]byte{jsoncdc.MustEncode(cadence.Address(userAddress))},
		)
		assert.EqualValues(t, CadenceUFix64("50.0"), userBalance)

		// Assert that total supply is correct
		supply := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetSupplyScript(fungibleAddr, kibbleAddr),
			nil,
		)
		assert.EqualValues(t, CadenceUFix64("50.0"), supply)
	})
}

func TestKibbleTransfers(t *testing.T) {
	b := newEmulator()

	fungibleAddr, kibbleAddr, kibbleSigner := KibbleDeployContracts(b, t)

	userAddress, _ := KibbleCreateAccount(t, b, fungibleAddr, kibbleAddr)

	KibbleMint(t, b, fungibleAddr, kibbleAddr, kibbleSigner, kibbleAddr, "1000.0", false)

	t.Run("Should not be able to withdraw more than the balance of the vault", func(t *testing.T) {
		tx := flow.NewTransaction().
			SetScript(kibbleGenerateTransferVaultScript(fungibleAddr, kibbleAddr)).
			SetGasLimit(100).
			SetProposalKey(b.ServiceKey().Address, b.ServiceKey().Index, b.ServiceKey().SequenceNumber).
			SetPayer(b.ServiceKey().Address).
			AddAuthorizer(kibbleAddr)

		_ = tx.AddArgument(CadenceUFix64("30000.0"))
		_ = tx.AddArgument(cadence.NewAddress(userAddress))

		signAndSubmit(
			t, b, tx,
			[]flow.Address{b.ServiceKey().Address, kibbleAddr},
			[]crypto.Signer{b.ServiceKey().Signer(), kibbleSigner},
			true,
		)

		// Assert that vault balances are correct

		kibbleBalance := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr),
			[][]byte{jsoncdc.MustEncode(cadence.Address(kibbleAddr))},
		)
		assert.EqualValues(t, CadenceUFix64("1000.0"), kibbleBalance)

		userBalance := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr),
			[][]byte{jsoncdc.MustEncode(cadence.Address(userAddress))},
		)
		assert.EqualValues(t, CadenceUFix64("0.0"), userBalance)
	})

	t.Run("Should be able to withdraw and deposit tokens from a vault", func(t *testing.T) {
		tx := flow.NewTransaction().
			SetScript(kibbleGenerateTransferVaultScript(fungibleAddr, kibbleAddr)).
			SetGasLimit(100).
			SetProposalKey(b.ServiceKey().Address, b.ServiceKey().Index, b.ServiceKey().SequenceNumber).
			SetPayer(b.ServiceKey().Address).
			AddAuthorizer(kibbleAddr)

		_ = tx.AddArgument(CadenceUFix64("300.0"))
		_ = tx.AddArgument(cadence.NewAddress(userAddress))

		signAndSubmit(
			t, b, tx,
			[]flow.Address{b.ServiceKey().Address, kibbleAddr},
			[]crypto.Signer{b.ServiceKey().Signer(), kibbleSigner},
			false,
		)

		// Assert that vault balances are correct

		kibbleBalance := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr),
			[][]byte{jsoncdc.MustEncode(cadence.Address(kibbleAddr))},
		)
		assert.EqualValues(t, CadenceUFix64("700.0"), kibbleBalance)

		userBalance := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr),
			[][]byte{jsoncdc.MustEncode(cadence.Address(userAddress))},
		)
		assert.EqualValues(t, CadenceUFix64("300.0"), userBalance)

		supply := executeScriptAndCheck(
			t, b,
			kibbleGenerateGetSupplyScript(fungibleAddr, kibbleAddr),
			nil,
		)
		assert.EqualValues(t, CadenceUFix64("1000.0"), supply)
	})
}

func kibbleReplaceAddressPlaceholders(code string, fungibleAddress, kibbleAddress string) []byte {
	return []byte(replaceImports(
		code,
		map[string]*regexp.Regexp{
			fungibleAddress: ftAddressPlaceholder,
			kibbleAddress:   kibbleAddressPlaceHolder,
		},
	))
}

func loadFungibleToken() []byte {
	return ft_contracts.FungibleToken()
}

func loadKibble(fungibleAddr flow.Address) []byte {
	return []byte(replaceImports(
		string(readFile(kibbleContractPath)),
		map[string]*regexp.Regexp{
			fungibleAddr.String(): ftAddressPlaceholder,
		},
	))
}

func kibbleGenerateGetSupplyScript(fungibleAddr, kibbleAddr flow.Address) []byte {
	return kibbleReplaceAddressPlaceholders(
		string(readFile(kibbleGetSupplyPath)),
		fungibleAddr.String(),
		kibbleAddr.String(),
	)
}

func kibbleGenerateGetBalanceScript(fungibleAddr, kibbleAddr flow.Address) []byte {
	return kibbleReplaceAddressPlaceholders(
		string(readFile(kibbleGetBalancePath)),
		fungibleAddr.String(),
		kibbleAddr.String(),
	)
}
func kibbleGenerateTransferVaultScript(fungibleAddr, kibbleAddr flow.Address) []byte {
	return kibbleReplaceAddressPlaceholders(
		string(readFile(kibbleTransferTokensPath)),
		fungibleAddr.String(),
		kibbleAddr.String(),
	)
}

func kibbleGenerateSetupKibbleAccountTransaction(fungibleAddr, kibbleAddr flow.Address) []byte {
	return kibbleReplaceAddressPlaceholders(
		string(readFile(kibbleSetupAccountPath)),
		fungibleAddr.String(),
		kibbleAddr.String(),
	)
}

func kibbleGenerateMintKibbleTransaction(fungibleAddr, kibbleAddr flow.Address) []byte {
	return kibbleReplaceAddressPlaceholders(
		string(readFile(kibbleMintTokensPath)),
		fungibleAddr.String(),
		kibbleAddr.String(),
	)
}

File: cadence/lib/js/test/kibble.test.js
import path from "path";
import { emulator, init, getAccountAddress, shallPass, shallResolve, shallRevert } from "flow-js-testing";

import {
	deployKibble,
	setupKibbleOnAccount,
	getKibbleBalance,
	getKibbleSupply,
	mintKibble,
	transferKibble,
} from "../src/kibble";

import { toUFix64, getRegistry } from "../src/common";

// We need to set timeout for a higher number, cause some transactions might take up some time
jest.setTimeout(30000);

describe("Kibble", () => {
	// Instantiate emulator and path to Cadence files
	beforeEach(async (done) => {
		const basePath = path.resolve(__dirname, "../../../");
		const port = 8080;
		init(basePath, port);
		await emulator.start(port, false);
		done();
	});

	// Stop emulator, so it could be restarted
	afterEach(async (done) => {
		await emulator.stop();
		done();
	});

	test("should have initialized supply field correctly", async () => {
		// Deploy contract
		await shallPass(deployKibble());

		await shallResolve(async () => {
			const supply = await getKibbleSupply();
			expect(supply).toBe(toUFix64(0));
		});
	});

	test("should be able to create empty Vault that doesn't affect supply", async () => {
		// Setup
		await deployKibble();
		const Alice = await getAccountAddress("Alice");
		await shallPass(setupKibbleOnAccount(Alice));

		await shallResolve(async () => {
			const supply = await getKibbleSupply();
			const aliceBalance = await getKibbleBalance(Alice);
			expect(supply).toBe(toUFix64(0));
			expect(aliceBalance).toBe(toUFix64(0));
		});
	});

	test("should not be able to mint zero tokens", async () => {
		// Setup
		await deployKibble();
		const Alice = await getAccountAddress("Alice");
		await setupKibbleOnAccount(Alice);

		// Mint instruction with amount equal to 0 shall be reverted
		const result = await shallRevert(mintKibble(Alice, toUFix64(0)));
	});

	test("Should mint tokens, deposit, and update balance and total supply", async () => {
		// Setup
		await deployKibble();
		const Alice = await getAccountAddress("Alice");
		await setupKibbleOnAccount(Alice);
		const amount = toUFix64(50);

		// Mint Kibble tokens for Alice
		await shallPass(mintKibble(Alice, amount));

		// Check Kibble total supply and Alice's balance
		await shallResolve(async () => {
			// Check Alice balance to equal amount
			const balance = await getKibbleBalance(Alice);
			expect(balance).toBe(amount);

			// Check Kibble supply to equal amount
			const supply = await getKibbleSupply();
			expect(supply).toBe(amount);
		});
	});

	test("should not be able to withdraw more than the balance of the Vault", async () => {
		// Setup
		await deployKibble();
		const Registry = await getRegistry();
		const Alice = await getAccountAddress("Alice");
		await setupKibbleOnAccount(Registry);
		await setupKibbleOnAccount(Alice);

		// Set amounts
		const amount = toUFix64(1000);
		const overflowAmount = toUFix64(30000);

		// Mint instruction shall resolve
		await shallResolve(mintKibble(Registry, amount));

		// Transaction shall revert
		await shallRevert(transferKibble(Registry, Alice, overflowAmount));

		// Balances shall be intact
		await shallResolve(async () => {
			const aliceBalance = await getKibbleBalance(Alice);
			expect(aliceBalance).toBe(toUFix64(0));

			const registryBalance = await getKibbleBalance(Registry);
			expect(registryBalance).toBe(amount);
		});
	});

	test("should be able to withdraw and deposit tokens from a Vault", async () => {
		await deployKibble();
		const Registry = await getRegistry();
		const Alice = await getAccountAddress("Alice");
		await setupKibbleOnAccount(Registry);
		await setupKibbleOnAccount(Alice);
		await mintKibble(Registry, toUFix64(1000));

		await shallPass(transferKibble(Registry, Alice, toUFix64(300)));

		await shallResolve(async () => {
			// Balances shall be updated
			const registryBalance = await getKibbleBalance(Registry);
			expect(registryBalance).toBe(toUFix64(700));

			const aliceBalance = await getKibbleBalance(Alice);
			expect(aliceBalance).toBe(toUFix64(300));

			const supply = await getKibbleSupply();
			expect(supply).toBe(toUFix64(1000));
		});
	});
});
